internal/kibana/slo.go

```diff
@@ -0,0 +1,477 @@
+				Type: indicatorType,
+				Params: slo.IndicatorPropertiesApmAvailabilityParams{
+					Service:         d.Get("indicator.0.params.0.service").(string),
+					Environment:     d.Get("indicator.0.params.0.environment").(string),
+					TransactionType: d.Get("indicator.0.params.0.transaction_type").(string),
+					TransactionName: d.Get("indicator.0.params.0.transaction_name").(string),
+					Filter:          getOrNilString("indicator.0.params.0.filter", d),
+					Index:           d.Get("indicator.0.params.0.index").(string),
+				},
+			},
+		}
+
+	case "sli.apm.transactionDuration":
+		indicator = slo.SloResponseIndicator{
+			IndicatorPropertiesApmLatency: &slo.IndicatorPropertiesApmLatency{
+				Type: indicatorType,
+				Params: slo.IndicatorPropertiesApmLatencyParams{
+					Service:         d.Get("indicator.0.params.0.service").(string),
+					Environment:     d.Get("indicator.0.params.0.environment").(string),
+					TransactionType: d.Get("indicator.0.params.0.transaction_type").(string),
+					TransactionName: d.Get("indicator.0.params.0.transaction_name").(string),
+					Filter:          getOrNilString("indicator.0.params.0.filter", d),
+					Index:           d.Get("indicator.0.params.0.index").(string),
+					Threshold:       float64(d.Get("indicator.0.params.0.threshold").(int)),
+				},
+			},
+		}
+
+	default:
+		return models.Slo{}, diag.Errorf("unknown indicator type %s", indicatorType)
+	}
+
+	timeWindow := slo.TimeWindow{
+		Type:     d.Get("time_window.0.type").(string),
+		Duration: d.Get("time_window.0.duration").(string),
+	}
+
+	objective := slo.Objective{
+		Target:          d.Get("objective.0.target").(float64),
+		TimesliceTarget: getOrNilFloat("objective.0.timeslice_target", d),
+		TimesliceWindow: getOrNilString("objective.0.timeslice_window", d),
+	}
+
+	var settings slo.Settings
+	if _, ok := d.GetOk("settings"); ok {
+		settings = slo.Settings{
+			SyncDelay: getOrNilString("settings.0.sync_delay", d),
+			Frequency: getOrNilString("settings.0.frequency", d),
+		}
+	}
+
+	slo := models.Slo{
+		Name:            d.Get("name").(string),
+		Description:     d.Get("description").(string),
+		Indicator:       indicator,
+		TimeWindow:      timeWindow,
+		BudgetingMethod: d.Get("budgeting_method").(string),
+		Objective:       objective,
+		Settings:        &settings,
+		SpaceID:         d.Get("space_id").(string),
+	}
+
+	return slo, diags
+}
+
+func resourceSloCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client, diags := clients.NewApiClient(d, meta)
+	if diags.HasError() {
+		return diags
+	}
+
+	slo, diags := getSloFromResourceData(d)
+	if diags.HasError() {
+		return diags
+	}
+
+	res, diags := kibana.CreateSlo(ctx, client, slo)
+
+	if diags.HasError() {
+		return diags
+	}
+
+	id, diags := client.ID(ctx, res.ID)
+	if diags.HasError() {
+		return diags
+	}
+
+	id.ResourceId = res.ID
+	d.SetId(id.String())
```

Comment:
Can you elaborate here? This is the same pattern I'm seeing everywhere in the codebase.

Should it instead be:
```go
	compId, diags := clients.CompositeIdFromStr(d.Id())
	if diags.HasError() {
		return diags
	}
	d.SetId(compId.String())
```

Reply:
When I did this, I got
```
        Error: Wrong resource ID.

          with elasticstack_kibana_slo.test_slo,
          on terraform_plugin_test.tf line 12, in resource "elasticstack_kibana_slo" "test_slo":
          12: resource "elasticstack_kibana_slo" "test_slo" {

        Resource ID must have following format: <cluster_uuid>/<resource identifier>
```

Reply:
Probably because you had an existing resource, You'll have to delete any resources before updating this to something like:

```
	id := &clients.CompositeId{ClusterId: rule.SpaceID, ResourceId: res.ID}
	d.SetId(id.String())
```

Reply:
Ok, cool. Will give this a whirl.